### PR TITLE
Feature/handle polymorphic

### PR DESCRIPTION
**Pull request by SereV94**
Main changes introduced:
- Allow **only one subtype of a polymorphic** at any nesting level within a structure or a collection
- Add support for **nullable polymorphic**
- Remove delegation from the instantiation of polymorphic serializers (**inherit directly from SurrogateSerializer** instead)

### CODE REVIEW

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt [hunk 1]
```diff
@@ -72,69 +89,121 @@ class ElementFactory(
...
                 val variantNamesLengths = polyDescriptors.map { it.serialName.length }.distinct()
                 if (variantNamesLengths.size != 1) {
                     throw SerdeError.VariablePolymorphicSerialName(descriptor)
                 }
                 dfQueue.prepend(variantNamesLengths.single())
+                // retrieve the base class of the polymorphic and cast it appropriately
+                val polyBaseClass: KClass<in Any> = descriptor.capturedKClass as? KClass<in Any>
+                    ?: throw SerdeError.NoPolymorphicBaseClass(descriptor.serialName)
+                // Polymorphic type consists of a string describing type and a structure.
+                val children = mutableListOf(
```
Comment (vnermolaev) [inline src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:154]: you could use some spacing in-between less related lines.

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt [hunk 2]
```diff
@@ -1,20 +1,62 @@
 package com.ing.serialization.bfl.serde.element
 
+import com.ing.serialization.bfl.serde.SerdeError
```
Comment (vnermolaev) [inline src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt:3]: I think it is better to have a separate `PolymorphicStructureElement`, now you have not only an extra field but a specific `decodePolymorphicNull` method which is quite meaty.

**Review by vnermolaev — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt [hunk 3]
```diff
@@ -1,9 +1,102 @@
...
+ * @throws SerdeError.DifferentPolymorphicImplementations when different implementations of the same polymorphic base
+ * type are encountered
+ */
+fun Element.merge(other: Element): Element {
+    require(this::class == other::class) { "Elements to be merged should be of the same type" }
+    return if (isNull || other.isNull) {
+        if (isNull) other.clone() else this.clone()
+    } else {
+        // Polymorphic type consists of a string describing type and a structure
+        if (isPolymorphic && inner.last().serialName != other.inner.last().serialName) {
```
[the remark this answers is not in the document]
Comment (vnermolaev) [reply src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt:74]: this can be scheduled for the next PR.

**Comment by vnermolaev** [comment]
Good work. Let's add some polish to make it shine :)

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt [hunk 4]
```diff
@@ -1,64 +1,81 @@
...
+                }
             }
 
             else -> error("${descriptor.serialName} is not supported")
         }
     }
 
-    @Suppress("ComplexMethod")
-    private fun fromType(descriptor: SerialDescriptor, parentName: String): Element {
+    @Suppress("ComplexMethod", "LongMethod", "NestedBlockDepth")
```
[the remark this answers is not in the document]
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:77]: Yes, it is... Maybe I could create separate functions for collections and polymorphic, since these 2 cases occupy most lines

File: src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt [hunk 5]
```diff
@@ -33,7 +34,17 @@ class FixedLengthStructureProcessor(
...
+        var schedulable = queue.first().expect<StructureElement>()
+
+        // during decoding we need to populate the inner placeholder StructureElement of a polymorphic
+        val parent = schedulable.parent
+        if (parent != null && parent.isPolymorphic && phase == Phase.DECODING) {
+            // populate the placeholder StructureElement
+            schedulable = parent.resolvePolymorphicChild(descriptor, schedulable.propertyName, serializersModule)
+            // remove the placeholder StructureElement from queue and add the populated version
+            removeNext()
+            queue.prepend(schedulable)
```
[the remark this answers is not in the document]
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt:46]: Because initially there is the inner placeholder element in the queue... So, after resolution we remove it from the queue (line 45) and push the resolved inner StructureElement in its place (line 46)

File: src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt [hunk 6]
```diff
@@ -1,9 +1,102 @@
...
+            throw SerdeError.DifferentPolymorphicImplementations(serialName)
+        }
+        inner = inner.mapIndexed { idx, child -> child.merge(other.inner[idx]) }.toMutableList()
+        this.clone()
+    }
+}
+
+/**
+ * Extension function used for resolving the inner StructureElement of a polymorphic when its actual type is available
```
[the remark this answers is not in the document]
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt:83]: Indeed, the docStrings need to be improved. What we do have at that point is the correct descriptor of the inner StructureElement of the polymorphic. I will update the description of the function

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt (hunk 2 above)
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt:3]: I am not sure that this option would improve code quality, because the variable `isPolymorphic` used currently to denote a polymorphic Element, is defined on `Element`, giving us a lot of flexibility in the declaration of functions operating on `Elements`. If we have a dedicated `PolymorphicStructureElement`, then the above flag is not needed, and explicit casting is needed before calling functions that exist only for `PolymorphicStructureElement`

Long story short, we can do it but I am not sure if it is indeed beneficial

File: src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt:74]: It's doable. Shall I include it in this PR or handle it in a separate one?

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt [hunk 7]
```diff
@@ -72,69 +89,121 @@ class ElementFactory(
...
+
+                // first try to treat data as list using reflection
+                val children = data?.convertToList()?.let {
+                    // in case of an empty collection, don't pass any data to a deeper parsing level
+                    if (it.isEmpty()) {
+                        return@let descriptor.elementDescriptors.map { innerDescriptor -> fromType(innerDescriptor, parentName) }.toMutableList()
+                    }
+                    // if the collection is not empty, it can be of either of List-like or Map-like type
+                    if (data is Map<*, *> || data is MutableMap<*, *>) {
+                        it.filterIsInstance<Pair<*, *>>().unzip().toList()
```
[the remark this answers is not in the document]
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:101]: I am not sure that I understand the question. Are you referring to doing this special handling for Maps and Lists within `convertToList` instead of `ElementFactory`?

File: src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (vnermolaev) [reply src/main/kotlin/com/ing/serialization/bfl/serde/Utils.kt:74]: Let's do it in the next one, there are other smaller improvements.

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (vnermolaev) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:101]: `is MutableMap<*, *> -> this.toList().unzip().toList()`
in `fun <T> T.convertToList()`

File: src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (vnermolaev) [reply src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt:46]: Ah, I see. Please add comment `Replace ...`

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt (hunk 2 above)
Comment (vnermolaev) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt:3]: well, `PolymorphicStructureElement` is still an `Element` you can pass it in any function accepting `Element` without extra cast.
Flag `isPolymorphic` for any `element: Element` is easily reproduced `element is PolymorphicStructureElement`.

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:101]: Indeed, this can be incorporated in `convertToList` and it would simplify the collection handling in ElementFactory. Nevertheless, `convertToList` would need to return a List of Lists in that case. If this is fine, I can easily make this change

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt (hunk 2 above)
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt:3]: Indeed, but consider the function `verifyResolvabilityOrThrow` defined on `Element`. This function is needed to see whether a nullable `Element` containing a polymorphic can be resolved. If we go with your suggestion, then we would refer to a subclass (`PolymorphicStructureElement`) within the base class (`Element`)

Comment (vnermolaev) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/StructureElement.kt:3]: I am not sure what the problem is.
```kotlin
fun verifyResolvabilityOrThrow(): Element {
        // if a null polymorphic has not been resolved in the parsing stage, an exception is thrown
        if (isNull && this is PolymorphicStructureElement) {
            throw SerdeError.NonResolvablePolymorphic(serialName)
        }
        // check also the children of the element for resolvability
        inner.forEach { it.verifyResolvabilityOrThrow() }
        return this
    }
```

File: src/main/kotlin/com/ing/serialization/bfl/serde/BinaryFixedLengthInputDecoder.kt [hunk 8]
```diff
@@ -80,14 +81,17 @@ class BinaryFixedLengthInputDecoder(
     override fun decodeNotNullMark() = input.readBoolean()
 
     override fun decodeNull(): Nothing? {
-        structureProcessor.removeNext().decodeNull(input)
+        structureProcessor.removeNext().let {
+            if (it is PolymorphicStructureElement) it.decodeNull(input, serializersModule)
```
Comment (vnermolaev) [inline src/main/kotlin/com/ing/serialization/bfl/serde/BinaryFixedLengthInputDecoder.kt:85]: is it reasonable to unify `decodeNull` for all kinds of elements? i.e., `decodeNull(input: DataInputStream, serializersModule: SerializersModule = EmptySerializersModule)`. In this case, we can simplify the flow a bit.

File: src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt [hunk 9]
```diff
@@ -33,7 +35,17 @@ class FixedLengthStructureProcessor(
...
         // TODO: add check if the struct on the stack coincides with the current descriptor.
-        val schedulable = queue.first().expect<StructureElement>()
+        var schedulable = queue.first().expect<StructureElement>()
+
+        // during decoding we need to populate the inner placeholder StructureElement of a polymorphic
+        val parent = schedulable.parent
+        if (phase == Phase.DECODING && parent != null && parent is PolymorphicStructureElement) {
+            // populate the placeholder StructureElement
+            schedulable = parent.resolvePolymorphicChild(descriptor, schedulable.propertyName, serializersModule)
+            // replace the placeholder StructureElement from queue with the populated version
```
Comment (vnermolaev) [inline src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt:45]: maybe we can still replace the following the two next lines with a devoted ext function `reschedule(schedulable)` on `ArrayDeque` which would replace the first. Then the code is more self-explanatory.

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt [hunk 10]
```diff
@@ -69,72 +84,153 @@ class ElementFactory(
...
+    private fun fromCollection(
+        descriptor: SerialDescriptor,
+        serialName: String,
+        parentName: String,
+        data: Any? = null
+    ): CollectionElement {
+        val requiredLength = dfQueue.removeFirstOrNull()
+            ?: throw SerdeError.InsufficientLengthData(descriptor, parentName)
+
+        // use reflection to treat non-null collection data
```
Comment (vnermolaev) [inline src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:132]: why is this comment here? seems an artefact from the prev version.

**Review by vnermolaev — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt (hunk 10 above)
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/element/ElementFactory.kt:132]: True, I will remove it

File: src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt (hunk 9 above)
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/FixedLengthStructureProcessor.kt:45]: We can. I will address it

File: src/main/kotlin/com/ing/serialization/bfl/serde/BinaryFixedLengthInputDecoder.kt (hunk 8 above)
Comment (SereV94) [reply src/main/kotlin/com/ing/serialization/bfl/serde/BinaryFixedLengthInputDecoder.kt:85]: Your suggestion would indeed simplify these lines, but I think that it would produce confusion when checking the definition of all the other elements apart from PolymorphicStructureElement, since serializersModule will be unused for them.

**Review by vnermolaev — APPROVED** [review]
